resources/type-system/subtyping.md

```diff
@@ -242,6 +242,140 @@ that the choice of common variable names avoid capture.  It is valid to choose
 the `Xi` or the `Yi` for `Zi` so long as capture is avoided*
 
 
+## Algorithmic subtyping: version 2
+
+This version of the algorithmic rules re-orders the rules to move all of the
+rules relevant to non-nullability as early as possible.  It also simplifies the
+algorithm somewhat by allowing specific rules to "fall through" to later rules
+when it is clearer to do so.  This presentation is intended to be equivalent to
+the above.  It is motivated by pragmatic concerns from the implementations.
+
+### Rules
+
+We say that a type `T` is **nullable** if `Null <: T`.  This is equivalent to
+the syntactic criterion that `T` is any of:
+  - `Null`
+  - `S?` for some `S`
+  - `S*` for some `S` where `S` is nullable
+  - `FutureOr<S>` for some `S` where `S` is nullable
+  - `dynamic`
+  - `void`
+
+We say that a type `T` is a **top** type if `T` is `dynamic`, `void`, `Object?`,
+or `FutureOr<S>` where `S` is a top type.
+
+**Note that these rules are designed to be read top down as an algorithm**
+
+We say that a type `T0` is a subtype of a type `T1` (written `T0 <: T1`) when:
+
+- **Reflexivity**: if `T0` and `T1` are the same type then `T0 <: T1` otherwise continue
+  - *Note that this check is necessary as the base case for primitive types, and
+    type variables but not for composite types.  In particular, algorithmically
+    a structural equality check is admissible, but not required
+    here. Pragmatically, non-constant time identity checks here are
+    counter-productive*
+
+- **Right Top**: if `T1` is a top type then `T0 <: T1` otherwise continue
+
+- **Right Legacy Top**: if `T1` is `S*` where `S` is a top type then `T0 <: T1`,
+  otherwise continue
+
+- **Left Top**: if `T0` is a top type then the subtyping does not hold (return
+  false)
+
+- **Left Legacy Top**: if `T0` is `S*` where `S` is a top type then the
+  subtyping does not hold (return false)
+
+- **Left Nullable 1**: if `T0` is `Null`, `Null*`, or `S0?` and `T1` is not
+  nullable then the subtyping does not hold (return false)
+
+- **Left Nullable 2**: otherwise, if `T0` is `S0?` then the subtyping holds iff
+  `S0 <: T1` (return true or false)
```

Comment:
I think ideally we wouldn't have to extract `S0` from `S0?`, because it requires a copy of `S0` which may be a moderately complex type like a function type.

That said, I think we are mostly worried about extracting `S0` from `S0*`, since legacy types will be much more common than nullable types for a long time. So unwrapping `S0?` may be totally doable.